Translated descriptions for policies, policy templates and policy template groups

alroniks opened the pull request:
### What does it do?
It replaces hardcoded strings in descriptions by keys and moves texts into lexicon files + some work to get it worked after moving.

Fixed several additional issues, found during the works:
- Fixed template group for ContextTemplate policy template
- Fixed list of core groups, templates, and policies, unified code for protecting them from deleting.
- Drastically simplified data-files that used for generating modx core package, also moving names to class constants to make the code more reliable
- Removed ability in-row editing of descriptions of policies and templates to avoid uncertainties when lexicon keys are used. In the full form descriptions updated and clear, but in the table, it is difficult to explain to the user what is going on.
- Added ability to change template group of created Policy Template. Remastered editing form.

![Policy Template_ AdministratorTemplate _ MODX Revolution](https://user-images.githubusercontent.com/303498/73435545-e9177380-4359-11ea-9a7d-4667724f3c32.png)

### Why is it needed?
It adds an ability to see descriptions of access policy templates and their groups on the user's native language (if translations exist). 

### Related issue(s)/PR(s)
Fixes #11360

Commit:
Translated descriptions for policies, policy templates and policy template groups

Reset translation files

alroniks commented:
I found out that policies also has not translations so I gonna fix them too.

Mark-H reviewed (dismissed):
Moar lexicons! Nice!

alroniks commented:
Too early @Mark-H :)
Need to fix processors and components for translatable names of policies.
It's mean that it shouldn't be merged for now (note for integrators). I will update the PR soon.

alroniks commented:
Tomorrow is 13th and it seems that I don't have time to finish this PR in time. Also, it will include additional big work for translators but we a limited in time. So I suggest moving to 3.x. I marked it for 3.x, later I will change the base of PR to 3.x and fix conflicts.

JoshuaLuckers commented:
@Alroniks are you still working on this?

alroniks commented:
@JoshuaLuckers Need to move it to 3.x and add some more translatuons but yes, I am working.

Mark-H commented:
Past the cut-off for new features and breaking changes in alpha1, so pushing forward. Maybe someone else could take the lead on getting this updated for 3.0.

alroniks commented:
@Mark-H @opengeek @JoshuaLuckers I finished, please, check.

CrazyBoy49z commented:
@Alroniks thanks

Ibochkarev commented:
@JoshuaLuckers please, change status - pr-state/review-needed

JoshuaLuckers commented:
@Alroniks would it be possible to fix the merge conflicts?

CrazyBoy49z commented:
> @Alroniks would it be possible to fix the merge conflicts?

Yes he fixed conflicts

alroniks commented:
@JoshuaLuckers I've fixed conflicts and did rebase as well.

alroniks commented:
@JoshuaLuckers @Jako @Mark-H @opengeek Does exist any chance that it will be reviewed?

JoshuaLuckers commented:
It appears there are some new conflicts that must be resolved. I'll review your PR tomorrow.

alroniks commented:
@JoshuaLuckers It is not a problem to fix these conflicts. But yes, it is annoying to fix them every time. That's why I am asking to check it and if it fits needs and requirements, merge it and go further.

JoshuaLuckers commented:
> @JoshuaLuckers It is not a problem to fix these conflicts. But yes, it is annoying to fix them every time. That's why I am asking to check it and if it fits needs and requirements, merge it and go further.

You are totally right! I'll check it out tomorrow!

JoshuaLuckers commented on _build/data/transport.core.accesspolicies.php:
```diff
@@ -9,136 +9,54 @@
-    'data' => '{"load":false,"list":false,"view":true}',
-    'lexicon' => 'permissions',
-], '', true, true);
+function jsonifyPermissions(array $permissions = []) {
+    $data = [];
+    foreach ($permissions as $key => $permission) {
+        if (is_string($key) && is_bool($permission)) {
+            $data[$key] = $permission;
+        } else {
+            $data[$permission] = true;
+        }
+    }
+    return json_encode($data);
+}
```
Wouldn't it make sense to have this functionality available to everyone who interacts with `modAccessPolicy::class`? Maybe it would be useful to do it automatically when setting the `data`? Or maybe when saving? Just some thoughts!

alroniks replied:
I agree, that it may be useful, but in this case, need to make this method as static to have the ability to use it in such scripts like this, that looks not good for a class that represents objects.
Move it to a separated helper also not looks good, so I would leave it as is before somebody will really need such functions in another place.

JoshuaLuckers replied:
Fair enough!

JoshuaLuckers commented on _build/data/transport.core.accesspolicies.php:
```diff
@@ -9,136 +9,54 @@
+    return json_encode($data);
+}
+
+$corePermissions = [
+    modAccessPolicy::POLICY_RESOURCE => ['add_children', 'create', 'copy', 'delete', 'list', 'load', 'move', 'publish', 'remove', 'save', 'steal_lock', 'undelete', 'unpublish', 'view'],
+    modAccessPolicy::POLICY_ADMINISTRATOR => ['about', 'access_permissions', 'actions', 'change_password', 'change_profile', 'charsets', 'class_map', 'components', 'content_types', 'countries', 'create', 'credits', 'customize_forms', 'dashboards', 'database', 'database_truncate', 'delete_category', 'delete_chunk', 'delete_context', 'delete_document', 'delete_eventlog', 'delete_plugin', 'delete_pr ...
+    modAccessPolicy::POLICY_LOAD_ONLY => ['load'],
+    modAccessPolicy::POLICY_LOAD_LIST_VIEW => ['load', 'list', 'view'],
+    modAccessPolicy::POLICY_OBJECT => ['load', 'list', 'view', 'save', 'remove'],
+    modAccessPolicy::POLICY_ELEMENT => ['add_children', 'create', 'delete', 'list', 'load', 'remove', 'save', 'view', 'copy'],
+    modAccessPolicy::POLICY_CONTENT_EDITOR => ['change_profile', 'class_map', 'countries', 'edit_document', 'frames', 'help', 'home', 'load', 'list', 'logout', 'menu_reports', 'menu_site', 'menu_support', 'menu_tools', 'menu_user', 'resource_duplicate', 'resource_tree', 'save_document', 'source_view', 'tree_show_resource_ids', 'view', 'view_document', 'view_template', 'new_document', 'delete_docu ...
+    modAccessPolicy::POLICY_MEDIA_SOURCE_ADMIN => ['create', 'copy', 'load', 'list', 'save', 'remove', 'view'],
+    modAccessPolicy::POLICY_MEDIA_SOURCE_USER => ['load', 'list', 'view'],
+    modAccessPolicy::POLICY_DEVELOPER => ['about', 'change_password', 'change_profile', 'charsets', 'class_map', 'components', 'content_types', 'countries', 'create', 'credits', 'customize_forms', 'dashboards', 'database', 'delete_category', 'delete_chunk', 'delete_context', 'delete_document', 'delete_eventlog', 'delete_plugin', 'delete_propertyset', 'delete_snippet', 'delete_template', 'delete_t ...
+    modAccessPolicy::POLICY_CONTEXT => ['load', 'list', 'view', 'save', 'remove', 'copy', 'view_unpublished'],
+    modAccessPolicy::POLICY_HIDDEN_NAMESPACE => ['load' => false, 'list' => false, 'view' => true],
+];
```
I'm curious as to why the core permissions are defined here but the policy names as constants in the object?
Furthermore I would go for a multiline formatting of the arrays to make it, in my opinion, more readable.

alroniks replied:
Because it is the only initial setup, after installation this list can be changed by users at any time, and having the default list does not makes sense. And in constants I stored the names, here I reuse these constant to better readability of code. The problem that we store all these permissions and policies in the database, and we can not fix it in the class.

JoshuaLuckers commented on _build/data/transport.core.accesspolicytemplategroups.php:
```diff
@@ -7,56 +7,24 @@
-]);
+foreach (modAccessPolicyTemplateGroup::getCoreGroups() as $index => $groupName) {
+    $group = $xpdo->newObject(modAccessPolicyTemplateGroup::class);
+    $group->fromArray(
+        [
+            'id' => $index + 1,
+            'name' => $groupName,
+            'description' => sprintf('policy_template_group_%s_desc', strtolower($groupName)),
+        ]
```
Isn't the `id` a primary key here that would require the parameter `setPrimaryKeys` set to `true`?
I'm just wondering what makes this different from the `modAccessPolicy` objects we create.

alroniks replied:
My solution based on the previous script (https://github.com/modxcms/revolution/blob/3.x/_build/data/transport.core.accesspolicytemplategroups.php), and this index guaranties that order will be always the same. Maybe I can simplify it more, I don't remember exactly already why I chose this solution.

JoshuaLuckers commented on core/src/Revolution/Processors/Security/Access/Policy/GetList.php:
```diff
@@ -118,15 +118,16 @@ public function beforeIteration(array $list)
+        $policy = $object->toArray();
+
+        $classes = ['pedit'];
+        if (!in_array($object->get('name'), modAccessPolicy::getCorePolicies(), true)) {
```
I'm considering to add a `isCorePolicy` method to the `modAccessPolicy*` models, what do you think?

alroniks replied:
Yes, it makes sense.

JoshuaLuckers commented on setup/includes/upgrades/common/3.0-policy-description.php:
```diff
@@ -0,0 +1,19 @@
+);
+
+foreach ($policies as $policy) {
+    $policy->set('description', sprintf(
+        'policy_%s_desc',
+        str_replace([',', ' '], ['', '_'], strtolower($policy->get('name')))
+    ));
```
What if I changed the description of one of the core policies to something custom? Would it be possible to check if the description equals the old "default" and only then update it?

The same question applies to the two other upgrade scripts.

alroniks replied:
It is pretty safe, because now in the database we store names as-is. And in constants, we also stored such values. So, if you changed something in the database, the following query just will not find such entries because it does not equal the requested values.
```
$policies = $modx->getIterator(
    modAccessPolicy::class,
    ['name:IN' => modAccessPolicy::getCorePolicies()]
);
```

JoshuaLuckers commented on core/lexicon/en/policy.inc.php:
```diff
@@ -29,7 +30,7 @@
 $_lang['policy_desc'] = 'Access policies are generic policies that restrict or enable certain actions with MODX.';
 $_lang['policy_desc_name'] = 'The name of the Access Policy';
-$_lang['policy_desc_description'] = 'Optional. A short description of the Access Policy';
+$_lang['policy_desc_description'] = 'Optional. A short description of the Access Policy. Also you might use lexicon keys here.';
```
Can we use all available lexicon keys or are they limited to the policy topic?

alroniks replied:
Based on language topics defined in controllers, so I would say that yes, mostly related to access control, but need to check because the manager loads a lot of topics and all of them will be cached.

JoshuaLuckers commented on core/lexicon/en/policy.inc.php:
```diff
@@ -55,10 +56,11 @@
 $_lang['policy_update'] = 'Update Policy';
 $_lang['policy_template'] = 'Policy Template';
-$_lang['policy_template.desc'] = 'A Policy Template defines which Permissions will show up in the Permissions grid when editing a specific Policy. You can add or remove specific Permissions from this template below. Note that removing a Permission from a Template will remove it from any Policies that use this Template.';
+$_lang['policy_template_desc'] = 'A Policy Template defines which Permissions will show up in the Permissions grid when editing a specific Policy. You can add or remove specific Permissions from this template below. Note that removing a Permission from a Template will remove it from any Policies that use this Template.';
```
If you're wondering (just like me 😅) what the change is: The `.` is replaced in the lexicon key!

alroniks replied:
It is done just to simplify logic in updater scripts. And a side effect, make lexicon keys more consistent. :)

JoshuaLuckers replied:
It's a good change, I just had a hard time spotting it haha.

JoshuaLuckers commented on core/src/Revolution/modAccessPolicy.php:
```diff
@@ -25,6 +25,42 @@
+    const POLICY_CONTEXT = 'Context';
+    const POLICY_HIDDEN_NAMESPACE = 'Hidden Namespace';
+
+    /**
+     * Returns list of core Policies
+     *
+     * @return array
+     */
+    public static function getCorePolicies()
+    {
+        return [
+            self::POLICY_RESOURCE,
+            self::POLICY_ADMINISTRATOR,
+            self::POLICY_LOAD_ONLY,
+            self::POLICY_LOAD_LIST_VIEW,
+            self::POLICY_OBJECT,
+            self::POLICY_ELEMENT ,
+            self::POLICY_CONTENT_EDITOR,
+            self::POLICY_MEDIA_SOURCE_ADMIN,
+            self::POLICY_MEDIA_SOURCE_USER,
+            self::POLICY_DEVELOPER,
+            self::POLICY_CONTEXT,
+            self::POLICY_HIDDEN_NAMESPACE,
+        ];
+    }
```
Since the core polices don't change, have you considered using a constant for the list of core policies instead of a static method?

alroniks replied:
I prefer method because it looks better in code, where it used, but yes, it is possible to move them to constant.
And another reason, method easier to override if needed (but I don't see any example when it can be required). :)

JoshuaLuckers reviewed (commented):
Great work! I left some general feedback and questions, nothing show stopping!

alroniks commented:
@JoshuaLuckers I have move ideas about improvements, but I prefer to stop and finish with it at first, to do not make a pull request too big (it is already pretty big).

JoshuaLuckers commented:
If you have time to resolve the conflicts I'll test it today or tomorrow and if everything works I'll approve it!

alroniks commented:
@JoshuaLuckers Conflicts are resolved, as well I added some improvements with methods for checking core items.

alroniks commented:
Besides conflicts (honestly, I already tired fix them every time), @opengeek, do you have anything to say against? 
@JoshuaLuckers Did you have chance to test it? Are all work as expected?

opengeek commented:
The only issue I have with this is it needs to be squashed into 1 commit and force pushed to the PR. In addition, you should never merge changes from upstream into a PR branch. If you need to resolve conflicts that show up after you submit the PR, you can rebase the changes on top of the latest from upstream.

opengeek commented:
Actually, I see now that you are editing the translated files directly in this PR. I'm not sure that won't completely break the integration with CrowdIn.

Jako commented:
Crowdin detects renamed/changed lexicon keys, if the text itself does not changes much. So the other language changes should be removed.

alroniks commented:
> Actually, I see now that you are editing the translated files directly in this PR. I'm not sure that won't completely break the integration with CrowdIn.

It shouldn't because only keys updated, not the strings.

opengeek commented:
> > Actually, I see now that you are editing the translated files directly in this PR. I'm not sure that won't completely break the integration with CrowdIn.
> 
> It shouldn't because only keys updated, not the strings.

It will. You need to remove the translated files from the commit as the changed key will be synced to CrowdIn, CrowdIn will recognize the key changed and will automatically update the translations accordingly.

alroniks commented:
@opengeek I squashed all into one commit.

alroniks commented:
@opengeek I forgot to say, but I also removed all changes from locales except English. Please, check and if all fine, let's merge it.

Ibochkarev commented:
@opengeek Can we check and accept this PR?

Mark-H commented on core/src/Revolution/Processors/Security/Access/Policy/GetList.php:
```diff
@@ -137,16 +134,35 @@ public function prepareRow(xPDOObject $object)
+     */
+    protected function prepareRowClasses(xPDOObject $object)
+    {
+        if ($object->isCorePolicy($object->get('name'))) {
```
Shouldn't this be a `if not isCorePolicy` check?

alroniks replied:
Yep, you are right. I fixed it. Also in templates as well.

Mark-H commented on manager/assets/modext/widgets/security/modx.grid.access.policy.js:
```diff
@@ -71,7 +71,10 @@ MODx.grid.AccessPolicy = function(config) {
             ,width: 375
-            ,editor: { xtype: 'textarea' }
+            ,renderer: function(value, metaData, record) {
+                return record['data']['description_trans'];
```
By defining a custom renderer, the standard renderer to encode html is no longer applied. Do we consider the translation lexicon safe or user input? Also repeated below in the policy template grid.

alroniks replied:
Added encoding to avoid potential XSS.

Mark-H commented:
This is looking preeeeettty darn good 👌

Commit:
chore: remove deleted (actually moved to one file) upgrade scripts

Commit:
fix: condition which allows delete entries

Commit:
fix: add html encoding for translations to avoid xss

Commit:
fix: missed translate for changed policy description

Commit:
feat: fallback to english in case translation does not exist in the target language

alroniks commented:
@Mark-H @opengeek @JoshuaLuckers I fixed issues and improved some other things, as well I fixed conflicts. It would be nice if you guys will find time to check this one. I know, it is a big pull request, but still atomic and brings only one new feature, which requires a lot of work though.

Mark-H reviewed (approved):
This is one of those changesets that needs extra testing during the merge (i.e. does the build work, altho I see no reason why it wouldn't) but 👍

Ibochkarev reviewed (approved):
LGTM 👍 

Thanks for the hard work! Need to fix conflicts and merged it

Commit:
Minified javascript
